Tool Components

Absolutionism opened the pull request:
### Description
This PR aims to add support for the upcoming Tool Components to Skript. Allowing users to modify components of an item and create new tool components with features such as: mining speed, tool rules, tool rule drops, tool rule mining speed, damage and many more.

This will be marked as draft due to Tool Components being ApiStatus.Experimental at this time.
Open to any syntax changes.

---
**Target Minecraft Versions:** any <!-- 'any' means all supported versions -->
**Requirements:** none <!-- Required plugins, server software... -->
**Related Issues:** none <!-- Links to related issues -->

Absolutionism committed:
Starter Commit

Absolutionism committed:
PR ready

Absolutionism committed:
remove "runtime"

Absolutionism committed:
Starter Commit

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/ToolRuleWrapper.java:
```diff
@@ -0,0 +1,72 @@
+public class ToolRuleWrapper implements ToolRule {
+
+	private Collection<Material> blocks = new ArrayList<>();
+	private boolean correct = false;
+	private Float speed = 5.0F;
```
these need to reflect default behavior of nullable, these values also feel magic number like where are they taken from?
```suggestion
	private Boolean correct;
	private Float speed;
```

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/ToolRuleWrapper.java:
```diff
@@ -0,0 +1,72 @@
+	private boolean correct = false;
+	private Float speed = 5.0F;
+
+	public ToolRuleWrapper() {}
+	public ToolRuleWrapper(Collection<Material> blocks, Float speed, boolean correct) {
```
```suggestion
	public ToolRuleWrapper() {}

	public ToolRuleWrapper(Collection<Material> blocks, Float speed, boolean correct) {
```

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/elements/CondToolRuleDrops.java:
```diff
@@ -0,0 +1,58 @@
+public class CondToolRuleDrops extends PropertyCondition<ToolRule> {
+
+	static {
+		Skript.registerCondition(CondToolRuleDrops.class, ConditionType.PROPERTY,
+			"[the] tool rule drops of %toolrules% (is|are) enabled",
+			"[the] tool rule drops of %toolrules% (is|are) disabled"
+		);
```
Any chance we can possibly get one that reflects the behavior more?
```applescript
loop tool rules of tool component of {_item}:
    loop-tool rule is correct for drops
    broadcast "%loop-tool rule% is correct for drops"
```

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/elements/EffSecCreateToolRule.java:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	static {
+		Skript.registerSection(EffSecCreateToolRule.class, "create [a] [new] tool rule and store (it|the result) in %object%");
```
Still not a fan of "store in" for this might be worth while to look into the ReturnHandler class to example
```applescript
create a new tool rule:
  return the tool rule
set {_toolrule} to last created tool rule
```

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/elements/EffSecCreateToolRule.java:
```diff
@@ -0,0 +1,111 @@
+		Skript.registerSection(EffSecCreateToolRule.class, "create [a] [new] tool rule and store (it|the result) in %object%");
+		EventValues.registerEventValue(ToolRuleEvent.class, ToolRule.class, new Getter<ToolRule, ToolRuleEvent>() {
+			@Override
+			public @Nullable ToolRule get(ToolRuleEvent event) {
+				return event.getToolRuleWrapper();
+			}
```
This is NonNull no?

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/elements/EffToolRuleDrops.java:
```diff
@@ -0,0 +1,62 @@
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (!getParser().isCurrentEvent(ToolRuleEvent.class)) {
```
What would happen if we tried changing the state outside of the section then? The rule should be modifiable no?
Especially since we're able to get references outside of it.

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/elements/ExprNewToolComp.java:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	@Override
+	protected ToolComponent @Nullable [] get(Event event) {
+		return new ToolComponent[]{(new ItemStack(Material.AIR)).getItemMeta().getTool()};
+	}
```
This should cause a stack trace, air has no itemmeta

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/elements/ExprNewToolComp.java:
```diff
@@ -0,0 +1,56 @@
+public class ExprNewToolComp extends SimpleExpression<ToolComponent> {
+
+	static {
+		Skript.registerExpression(ExprNewToolComp.class, ToolComponent.class, ExpressionType.SIMPLE,
```
I'm not sure how I feel about a method to create a tool component out of nothing. I'm personally against this I'll wait to see what other members say tho.

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/elements/ExprToolCompDamage.java:
```diff
@@ -0,0 +1,127 @@
+public class ExprToolCompDamage extends PropertyExpression<Object, Number> {
+
+	static {
+		Skript.registerExpression(ExprToolCompMiningSpeed.class, Number.class, ExpressionType.PROPERTY,
```
Confused on the random support of itemtype, itemstack and slot

TheAbsolutionism replied:
Allowing itemtypes, itemstacks and slots allows direct manipulation to the items rather than having it only be to a tool component then applying that tool component to multiple items. Especially if the multiple items have different data for their tool components.

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/elements/ExprToolCompDamage.java:
```diff
@@ -0,0 +1,127 @@
+})
+@RequiredPlugins("Minecraft 1.20.6+")
+@Since("INSERT VERSION")
+public class ExprToolCompDamage extends PropertyExpression<Object, Number> {
```
Is there a reason why we aren't using a simple property expression?

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/elements/ExprToolCompMiningSpeed.java:
```diff
@@ -0,0 +1,128 @@
+
+	@Override
+	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
+		return switch (mode) {
+			case SET, REMOVE, ADD -> CollectionUtils.array(Number.class);
+			default -> null;
```
Might be worth adding reset, the javadocs for the `getDefaultMiningSpeed` and `setDefaultMiningSpeed` method mentions `1.0` is the standard mining speed making me feel like that's more of a default.

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/elements/ExprToolCompMiningSpeed.java:
```diff
@@ -0,0 +1,128 @@
+})
+@RequiredPlugins("Minecraft 1.20.6+")
+@Since("INSERT VERSION")
+public class ExprToolCompMiningSpeed extends PropertyExpression<Object, Number> {
```
Same as before is there a reason this isn't using a `SimplePropertyExpression`

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/elements/ExprToolCompMiningSpeed.java:
```diff
@@ -0,0 +1,128 @@
+public class ExprToolCompMiningSpeed extends PropertyExpression<Object, Number> {
+
+	static {
+		Skript.registerExpression(ExprToolCompMiningSpeed.class, Number.class, ExpressionType.PROPERTY,
+			"[the] [default] mining speed of %itemstacks/itemtypes/slots/toolcomponents%");
```
I don't believe default should be optional here, otherwise I feel like it might be confused with `break speed` which accounts for potion effects (haste), enchantments (efficiency) and attributes

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/elements/ExprToolCompRules.java:
```diff
@@ -0,0 +1,138 @@
+
+	@Override
+	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
+		return switch (mode) {
+			case SET, DELETE, ADD, REMOVE -> CollectionUtils.array(ToolRuleWrapper.class);
+			default -> null;
```
Should this not be `ToolRule[].class`? Explicitly forcing the wrapper seems weird and prevents non skript implementations

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/elements/ExprToolCompRules.java:
```diff
@@ -0,0 +1,138 @@
+				current.addAll(Arrays.stream(finalRules).toList());
+				component.setRules(current);
+			};
+			case REMOVE -> component -> {
+				List<ToolRule> current = component.getRules();
+				current.removeAll(Arrays.stream(finalRules).toList());
+				component.setRules(current);
```
the `ToolComponent` class has a `removeRule` method why is this not being used?

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/elements/ExprToolCompRules.java:
```diff
@@ -0,0 +1,138 @@
+			case SET -> component -> {
+				component.setRules(Arrays.stream(finalRules).toList());
+			};
+			case DELETE -> component -> {
+				component.setRules(new ArrayList<>());
+			};
```
Might be easier to do `component.getRules().clear()`
*edit: docs do say it's mutability of the list isn't guaranteed, I still think this is better than a full overwrite which could possibly poorly interact with other plugins if the same list was cached and not retrieved*

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/elements/ExprToolRuleBlocks.java:
```diff
@@ -0,0 +1,97 @@
+
+	static {
+		Skript.registerExpression(ExprToolRuleBlocks.class, ItemType.class, ExpressionType.PROPERTY,
+			"[the] tool rule blocks [of %toolrules%]");
```
Can we maybe use `block types` over `blocks` prevents an expectation of blocks this will also work better in the future if minecraft ever decides to expand this from just material ids and into blockdata similar to the `can place on` and `can destroy` components

Just fine to say no due to that atm not seemingly being planned

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/elements/ExprToolRuleBlocks.java:
```diff
@@ -0,0 +1,97 @@
+
+	@Override
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
+		if (!(event instanceof ToolRuleEvent toolRuleEvent))
+			return;
+
```
Why do we prevent modifications of toolrules outside of the creation section?

Fusezion commented on src/main/java/org/skriptlang/skript/bukkit/toolcomponent/elements/ExprToolRuleSpeed.java:
```diff
@@ -0,0 +1,90 @@
+})
+@RequiredPlugins("Minecraft 1.20.6+")
+@Since("INSERT VERSION")
+public class ExprToolRuleSpeed extends PropertyExpression<ToolRule, Number> {
```
I wonder if it's worth using `Float` instead of `Number` here, I believe skriptlang has said in the past expected types should reflect the return type only time it shouldn't is in syntax that should be kept as `%number%`

TheAbsolutionism replied:
I'm not entirely sure, see what the team says,

Efnilite replied:
if this always returns a float, you should make it a float

Fusezion reviewed (changes requested):
Here's 19 reviews to eventually be completed or ignored idk

Absolutionism committed:
Changes

sovdeeth committed:
Merge branch 'dev/feature' into dev/ToolComponents

sovdeeth committed:
Merge branch 'dev/feature' into dev/EquippableComponents

Absolutionism committed:
Partial Changes

Absolutionism committed:
Merge branch 'dev/ToolComponents' of https://github.com/TheAbsolutionism/SkriptTest into dev/ToolComponents

Absolutionism committed:
Merge branch 'dev/feature' into dev/ToolComponents

Absolutionism committed:
Merge branch 'dev/feature' into dev/EquippableComponents

Absolutionism committed:
Update

Absolutionism committed:
Update

Absolutionism committed:
Merge branch 'dev/feature' into dev/EquippableComponents

Absolutionism committed:
Merge branch 'dev/feature' into dev/ToolComponents

Absolutionism committed:
Merge remote-tracking branch 'upstream/dev/feature' into dev/EquippableComponents

# Conflicts:
#	src/main/java/ch/njol/skript/Skript.java
#	src/main/resources/lang/default.lang

Moderocky reviewed (approved):
I think this looks good. Well done.

Moderocky committed:
Merge branch 'dev/feature' into dev/ToolComponents

Absolutionism committed:
Merge branch 'dev/feature' into dev/ToolComponents

Absolutionism committed:
Merge remote-tracking branch 'upstream/dev/feature' into dev/EquippableComponents

# Conflicts:
#	src/main/java/ch/njol/skript/util/slot/EquipmentSlot.java

Absolutionism committed:
Merge branch 'dev/feature' into dev/EquippableComponents

Absolutionism committed:
Merge remote-tracking branch 'upstream/dev/feature' into dev/EquippableComponents

# Conflicts:
#	src/main/java/ch/njol/skript/Skript.java
#	src/main/java/ch/njol/skript/bukkitutil/SoundUtils.java
#	src/main/resources/lang/default.lang

Absolutionism committed:
Merge remote-tracking branch 'upstream/dev/feature' into dev/ToolComponents

# Conflicts:
#	src/main/java/ch/njol/skript/Skript.java

Absolutionism committed:
Update

Absolutionism committed:
Merge branch 'dev/feature' into dev/EquippableComponents

sovdeeth commented:
this looks good, but I would like to spend some more time investigating more cohesive ways to implement all the item components so I think this would be best for 2.11. Components are a major requirement to sunset aliases and something I'd like to have fully implemented for 2.11 rather than rushing into 2.10.

TheAbsolutionism commented:
> this looks good, but I would like to spend some more time investigating more cohesive ways to implement all the item components so I think this would be best for 2.11. Components are a major requirement to sunset aliases and something I'd like to have fully implemented for 2.11 rather than rushing into 2.10.

That's fine, I never expected this to be merged for 2.10 because Tool Components (Same with Equippable Components) are still experimental. Was just setting up a base at the very least.

Absolutionism committed:
Merge remote-tracking branch 'upstream/dev/feature' into dev/EquippableComponents

# Conflicts:
#	src/main/java/ch/njol/skript/Skript.java
#	src/main/java/ch/njol/skript/util/slot/EquipmentSlot.java
#	src/main/resources/lang/default.lang

Absolutionism committed:
Overhaul

Absolutionism committed:
Update CondEquipCompInteract.java

Absolutionism committed:
Requested Changes

Absolutionism committed:
Partial Changes

Absolutionism committed:
Merge remote-tracking branch 'upstream/dev/feature' into dev/EquippableComponents

# Conflicts:
#	src/main/resources/lang/default.lang

Absolutionism committed:
Update EquippableComponents.sk

Absolutionism committed:
revert

Absolutionism committed:
Merge branch 'dev/feature' into dev/EquippableComponents

Absolutionism committed:
Record Usage

Absolutionism committed:
Partial Changes

Absolutionism committed:
Merge remote-tracking branch 'upstream/dev/feature' into dev/EquippableComponents

# Conflicts:
#	src/main/java/ch/njol/skript/registrations/Feature.java
#	src/main/resources/lang/default.lang

Absolutionism committed:
More Changes

Absolutionism committed:
Merge remote-tracking branch 'upstream/dev/feature' into dev/ToolComponents

# Conflicts:
#	src/main/java/ch/njol/skript/Skript.java
#	src/main/resources/lang/default.lang

Absolutionism committed:
Merge branch 'dev/EquippableComponents' into dev/ToolComponents

# Conflicts:
#	src/main/java/ch/njol/skript/Skript.java
#	src/main/resources/lang/default.lang

Absolutionism committed:
Overhaul

Absolutionism committed:
Changes

Absolutionism committed:
Merge branch 'dev/EquippableComponents' into dev/ToolComponents

# Conflicts:
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/ItemComponentModule.java

Absolutionism committed:
Docs Update

Absolutionism committed:
Requested Changes

Absolutionism committed:
Update EquippableComponents.sk

Absolutionism committed:
Update EquippableComponents.sk

Absolutionism committed:
Adooblus Changes

Absolutionism committed:
Merge remote-tracking branch 'upstream/dev/feature' into dev/EquippableComponents

Absolutionism committed:
Update EquippableExperiment.java

Absolutionism committed:
Update Experiment Class  Name

Absolutionism committed:
Requested Changes

Absolutionism committed:
Cloneable Changes

Absolutionism committed:
Update EquippableModule.java

Absolutionism committed:
Update ItemComponentModule.java

Absolutionism committed:
Merge remote-tracking branch 'upstream/dev/feature' into dev/EquippableComponents

# Conflicts:
#	src/main/java/ch/njol/skript/Skript.java
#	src/main/java/ch/njol/skript/registrations/Feature.java
#	src/main/resources/lang/default.lang

Absolutionism committed:
Partial Update

Experiencing some issues, need to see if incorrect merge

Absolutionism committed:
Docs Update

Absolutionism committed:
Missing Import

Absolutionism committed:
More Elements

Absolutionism committed:
Docs Fix

Absolutionism committed:
Fix ItemType

Absolutionism committed:
Revert Misc Changed Files

Absolutionism committed:
Source Retainability

Absolutionism committed:
Update EquippableComponents.sk

Absolutionism committed:
Update ExprEquippableComponent.java

Absolutionism committed:
Requested Changes

Absolutionism committed:
Merge branch 'dev/EquippableComponents' into dev/ToolComponents

# Conflicts:
#	src/main/java/ch/njol/skript/registrations/Feature.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/ItemComponentModule.java

Absolutionism committed:
Partial Update - Broken

Absolutionism committed:
Merge remote-tracking branch 'upstream/dev/feature' into dev/ToolComponents

Absolutionism committed:
Fix Test

Absolutionism committed:
Javadocs

Absolutionism committed:
# Conflicts:
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/ComponentUtils.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/ComponentWrapper.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/EquippableWrapper.java

Absolutionism committed:
Update

Absolutionism committed:
Update ExprSecBlankToolComp.java

Absolutionism committed:
Merge dev/feature

Absolutionism committed:
Merge branch 'dev/feature' into dev/ToolComponents

Absolutionism committed:
Merge branch 'dev/feature' into dev/ToolComponents

Absolutionism committed:
Merge remote-tracking branch 'upstream/dev/feature' into dev/ToolComponents

# Conflicts:
#	src/main/java/ch/njol/skript/Skript.java
#	src/main/java/ch/njol/skript/registrations/Feature.java
#	src/main/java/ch/njol/skript/util/ItemSource.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/ComponentUtils.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/ComponentWrapper.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/ItemComponentModule.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/EquippableModule.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/EquippableWrapper.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/CondEquipCompDamage.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/CondEquipCompDispensable.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/CondEquipCompInteract.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/CondEquipCompShearable.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/CondEquipCompSwapEquipment.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/EffEquipCompDamageable.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/EffEquipCompDispensable.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/EffEquipCompInteract.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/EffEquipCompShearable.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/EffEquipCompSwapEquipment.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/ExprEquipCompCameraOverlay.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/ExprEquipCompEntities.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/ExprEquipCompEquipSound.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/ExprEquipCompModel.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/ExprEquipCompShearSound.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/ExprEquipCompSlot.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/ExprEquippableComponent.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/equippable/elements/ExprSecBlankEquipComp.java
#	src/main/java/org/skriptlang/skript/bukkit/itemcomponents/generic/ExprItemCompCopy.java
#	src/main/resources/lang/default.lang

Absolutionism committed:
Update

APickledWalrus commented on src/main/java/org/skriptlang/skript/bukkit/itemcomponents/tool/elements/CondToolRuleDrops.java:
```diff
@@ -0,0 +1,72 @@
+		Boolean correct = wrapper.getRule().correctForDrops().toBoolean();
+		if (correct != null)
+			return correct;
+		return false;
```
This should have the same behavior

```suggestion
		return wrapper.getRule().correctForDrops().toBooleanOrElse(false);
```

APickledWalrus commented on src/main/java/org/skriptlang/skript/bukkit/itemcomponents/tool/elements/CondToolRuleDrops.java:
```diff
@@ -0,0 +1,72 @@
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		SyntaxStringBuilder builder = new SyntaxStringBuilder(event, debug);
+		builder.append("the tool rule drops of", getExpr(), "are");
```
Maybe `getExpr().isSingle()` check for `is`/`are`? 🙂

APickledWalrus commented on src/main/java/org/skriptlang/skript/bukkit/itemcomponents/tool/elements/ExprToolCompDamage.java:
```diff
@@ -0,0 +1,80 @@
+		for (ToolWrapper wrapper : getExpr().getArray(event)) {
+			int newDamage;
+			if (mode == ChangeMode.ADD) {
+				newDamage = Math2.fit(0, wrapper.getComponent().damagePerBlock() + damage, Integer.MAX_VALUE);
```
I think this isn't ideal behavior when an overflow occurs. Like if a Skript user adds a really huge number, we probably just want to set the value to the maximum. Currently, an overflow would occur resulting in `newDamage` being `0`.

APickledWalrus commented on src/main/java/org/skriptlang/skript/bukkit/itemcomponents/tool/elements/ExprToolCompRules.java:
```diff
@@ -0,0 +1,112 @@
+	@Override
+	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
+		return switch (mode) {
+			case SET, DELETE, ADD, REMOVE -> CollectionUtils.array(ToolRuleWrapper[].class);
```
RESET might be okay here to function the same as DELETE

APickledWalrus reviewed (changes requested):
Looking really good overall. Here's what I found.

Absolutionism commented on src/main/java/org/skriptlang/skript/bukkit/itemcomponents/tool/elements/ExprToolRuleSpeed.java:
```diff
@@ -0,0 +1,102 @@
+	@Override
+	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
+		return switch (mode) {
+			case SET, DELETE, REMOVE, ADD -> CollectionUtils.array(Number.class);
```
I can make it where reset changes it to 1

Absolutionism committed:
Merge branch 'dev/feature' into dev/ToolComponents

Absolutionism committed:
Pickle's Changes

hotpoc commented:
i know i have no say in this whatsoever but i feel like it would be more user friendly to do:
`set mining speed (of|for) %blocks% for %item% to 50`
and then an expression or something like
`set drops (of|for) %blocks% for %item% to false`
